coco/state.py

```diff
@@ -364,14 +415,119 @@ def _load_default_state(self):
         for path, file in self.default_state_files.items():
             self.read_from_file(path, file)
 
-    async def process_post(self, request: dict = None):
+    def saved_state_exists(self, name: str):
+        """Check if a saved state with a given name exists."""
+        if name in self._saved_states:
+            return True
+        else:
+            return False
+
+    async def reset_state(self, request: dict = None):
         """
         Process the POST request to reset the state.
 
         Clear the internal state and re-load YAML files to restore default state.
         """
-        # back up excluded paths
-        excluded = dict()
+        excluded = self._backup_excluded_paths()
+
+        # Reset persistent state
+        with self._storage.update():
+            self._storage.state = dict()
+        self._load_default_state()
+
+        self._recover_excluded_paths(excluded)
+
+    async def save_state(self, request: dict = None):
+        """
+        Process the POST request to save (backup) the state.
+
+        The request dictionary should contain an item with key "name" that holds a string with the
+        name of the saved state.
+        """
+        # get request parameters
+        name = request.get("name", "backup")
+        if name == self._name_active_state:
+            raise InvalidUsage(
+                f"Can't use {self._name_active_state} for saved state. This name is reserved. "
+                f"Choose something else."
+            )
+        overwrite = request.get("overwrite", False)
+
+        # only overwrite an existing state if requested explicitly
+        if self.saved_state_exists(name):
+            if not overwrite:
+                raise InvalidUsage(
```

Comment:
Is it possible to return a Failed Request instead of raising an exception? Or is raising an exception the appropriate behaviour?

Reply:
This question applies to all of these requests.

Reply:
`worker.py` will catch the exception and build a HTTP response with a matching status code.

Reply:
Awesome